File: frontend/apps/ios/Pantopus/Core/Design/Components/LegalTOCCard.swift
//
//  LegalTOCCard.swift
//  Pantopus
//
//  A19 legal scaffold — a collapsible "Jump to section" table of contents.
//  The header toggles open/closed; each expanded row is a numbered section
//  link. State is owned by the host screen (it animates the toggle and
//  scrolls to the chosen section); the card itself is stateless.
//

import SwiftUI

/// A collapsible table-of-contents card. Bind `isOpen` + `onToggle` for the
/// header chevron, `items` for the section titles, and `onJump` for a row tap.
/// `onJump`'s index is the 0-based position in `items`; the matching heading
/// is `LegalSection(number: index + 1)`.
public struct LegalTOCCard: View {
    private let items: [String]
    private let isOpen: Bool
    private let onToggle: () -> Void
    private let onJump: (Int) -> Void

    public init(
        items: [String],
        isOpen: Bool,
        onToggle: @escaping () -> Void,
        onJump: @escaping (Int) -> Void
    ) {
        self.items = items
        self.isOpen = isOpen
        self.onToggle = onToggle
        self.onJump = onJump
    }

    public var body: some View {
        VStack(spacing: Spacing.s0) {
            header
            if isOpen {
                hairline
                rows
            }
        }
        .background(Theme.Color.appSurface)
        .clipShape(RoundedRectangle(cornerRadius: Radii.lg, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: Radii.lg, style: .continuous)
                .strokeBorder(Theme.Color.appBorder, lineWidth: 1)
        )
        .shadow(color: Theme.Color.appText.opacity(0.04), radius: 2, x: 0, y: 1)
        .accessibilityIdentifier("legalTOCCard")
    }

    // MARK: - Header

    private var header: some View {
        Button(action: onToggle) {
            HStack(spacing: Spacing.s2) {
                Icon(.list, size: 14, strokeWidth: 2.2, color: Theme.Color.primary600)
                Text("Jump to section")
                    .font(.system(size: 13, weight: .bold))
                    .foregroundColor(Theme.Color.appText)
                Spacer(minLength: Spacing.s2)
                if !isOpen {
                    Text("\(items.count) sections")
                        .font(.system(size: 11, weight: .medium))
                        .foregroundColor(Theme.Color.appTextSecondary)
                }
                Icon(
                    isOpen ? .chevronUp : .chevronDown,
                    size: 16,
                    color: Theme.Color.appTextSecondary
                )
            }
            .padding(.horizontal, 14)
            .padding(.vertical, Spacing.s3)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("legalTOCCard_toggle")
        .accessibilityLabel("Jump to section")
        .accessibilityValue("\(items.count) sections, \(isOpen ? "expanded" : "collapsed")")
        .accessibilityHint(isOpen ? "Collapses the table of contents" : "Expands the table of contents")
    }

    // MARK: - Rows

    private var rows: some View {
        VStack(spacing: Spacing.s0) {
            ForEach(Array(items.enumerated()), id: \.offset) { index, title in
                Button(action: { onJump(index) }, label: {
                    row(index: index, title: title)
                })
                .buttonStyle(.plain)
                .accessibilityIdentifier("legalTOCCard_row_\(index)")
                .accessibilityLabel(title)
                .accessibilityHint("Jumps to section \(index + 1)")

                if index < items.count - 1 {
                    hairline
                }
            }
        }
        .padding(.top, Spacing.s1)
        .padding(.bottom, 6)
    }

    private func row(index: Int, title: String) -> some View {
        HStack(spacing: 10) {
            Text(String(format: "%02d", index + 1))
                .font(.system(size: 10.5, weight: .bold))
                .foregroundColor(Theme.Color.primary700)
                .frame(width: 22, height: 22)
                .background(
                    RoundedRectangle(cornerRadius: Radii.sm, style: .continuous)
                        .fill(Theme.Color.primary50)
                )
            Text(title)
                .font(.system(size: 12.5, weight: .medium))
                .foregroundColor(Theme.Color.appTextStrong)
                .frame(maxWidth: .infinity, alignment: .leading)
            Icon(.chevronRight, size: 13, strokeWidth: 2, color: Theme.Color.appTextMuted)
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 9)
        .contentShape(Rectangle())
    }

    private var hairline: some View {
        Rectangle()
            .fill(Theme.Color.appBorderSubtle)
            .frame(height: 1)
    }
}

#Preview("Expanded") {
    LegalTOCCard(
        items: [
            "Overview", "Information we collect", "How we use it",
            "Identity pillars & privacy", "Sharing & disclosure"
        ],
        isOpen: true,
        onToggle: {},
        onJump: { _ in }
    )
    .padding(Spacing.s5)
    .background(Theme.Color.appSurface)
}

#Preview("Collapsed") {
    LegalTOCCard(
        items: Array(repeating: "Section", count: 10),
        isOpen: false,
        onToggle: {},
        onJump: { _ in }
    )
    .padding(Spacing.s5)
    .background(Theme.Color.appSurface)
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/components/LegalTOCCard.kt
@file:Suppress("MagicNumber", "LongMethod", "UnusedPrivateMember")

package app.pantopus.android.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusElevations
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing
import app.pantopus.android.ui.theme.pantopusShadow

/**
 * A19 legal scaffold — a collapsible "Jump to section" table of contents.
 * The header toggles open/closed; each expanded row is a numbered section
 * link. State is owned by the host screen (it animates the toggle and scrolls
 * to the chosen section); the card itself is stateless.
 *
 * @param items Section titles, in order.
 * @param isOpen Whether the section list is expanded.
 * @param onToggle Header tap — flip [isOpen].
 * @param onJump Row tap — the 0-based index in [items] (matching section is
 *     `LegalSection(number = index + 1)`).
 */
@Composable
fun LegalTOCCard(
    items: List<String>,
    isOpen: Boolean,
    onToggle: () -> Unit,
    onJump: (Int) -> Unit,
    modifier: Modifier = Modifier,
) {
    val shape = RoundedCornerShape(Radii.lg)
    Column(
        modifier =
            modifier
                .fillMaxWidth()
                .pantopusShadow(PantopusElevations.sm, shape)
                .clip(shape)
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, shape)
                .testTag("legalTOCCard"),
    ) {
        Row(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .clickable(onClick = onToggle)
                    .padding(horizontal = 14.dp, vertical = Spacing.s3)
                    .testTag("legalTOCCard_toggle"),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
        ) {
            PantopusIconImage(
                icon = PantopusIcon.List,
                contentDescription = null,
                size = 14.dp,
                strokeWidth = 2.2f,
                tint = PantopusColors.primary600,
            )
            Text(
                text = "Jump to section",
                fontSize = 13.sp,
                fontWeight = FontWeight.Bold,
                color = PantopusColors.appText,
            )
            Spacer(modifier = Modifier.weight(1f))
            if (!isOpen) {
                Text(
                    text = "${items.size} sections",
                    fontSize = 11.sp,
                    fontWeight = FontWeight.Medium,
                    color = PantopusColors.appTextSecondary,
                )
            }
            PantopusIconImage(
                icon = if (isOpen) PantopusIcon.ChevronUp else PantopusIcon.ChevronDown,
                contentDescription = null,
                size = 16.dp,
                tint = PantopusColors.appTextSecondary,
            )
        }

        if (isOpen) {
            TocHairline()
            Column(modifier = Modifier.padding(top = Spacing.s1, bottom = 6.dp)) {
                items.forEachIndexed { index, title ->
                    TocRow(index = index, title = title, onJump = onJump)
                    if (index < items.size - 1) {
                        TocHairline()
                    }
                }
            }
        }
    }
}

@Composable
private fun TocRow(
    index: Int,
    title: String,
    onJump: (Int) -> Unit,
) {
    Row(
        modifier =
            Modifier
                .fillMaxWidth()
                .clickable { onJump(index) }
                .padding(horizontal = 14.dp, vertical = 9.dp)
                .testTag("legalTOCCard_row_$index"),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Box(
            modifier =
                Modifier
                    .size(22.dp)
                    .clip(RoundedCornerShape(Radii.sm))
                    .background(PantopusColors.primary50),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                text = (index + 1).toString().padStart(2, '0'),
                fontSize = 10.5.sp,
                fontWeight = FontWeight.Bold,
                color = PantopusColors.primary700,
            )
        }
        Text(
            text = title,
            fontSize = 12.5.sp,
            fontWeight = FontWeight.Medium,
            color = PantopusColors.appTextStrong,
            modifier = Modifier.weight(1f),
        )
        PantopusIconImage(
            icon = PantopusIcon.ChevronRight,
            contentDescription = null,
            size = 13.dp,
            strokeWidth = 2f,
            tint = PantopusColors.appTextMuted,
        )
    }
}

@Composable
private fun TocHairline() {
    Box(
        modifier =
            Modifier
                .fillMaxWidth()
                .height(1.dp)
                .background(PantopusColors.appBorderSubtle),
    )
}

@Preview(showBackground = true, widthDp = 360)
@Composable
private fun LegalTOCCardExpandedPreview() {
    Column(modifier = Modifier.padding(Spacing.s5)) {
        LegalTOCCard(
            items =
                listOf(
                    "Overview",
                    "Information we collect",
                    "How we use it",
                    "Identity pillars & privacy",
                    "Sharing & disclosure",
                ),
            isOpen = true,
            onToggle = {},
            onJump = {},
        )
    }
}

@Preview(showBackground = true, widthDp = 360)
@Composable
private fun LegalTOCCardCollapsedPreview() {
    Column(modifier = Modifier.padding(Spacing.s5)) {
        LegalTOCCard(
            items = List(10) { "Section" },
            isOpen = false,
            onToggle = {},
            onJump = {},
        )
    }
}
